File: java/src/main/java/com/tramli/MermaidGenerator.java
package com.tramli;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashSet;
import java.util.Set;

/**
 * Generates Mermaid stateDiagram-v2 from FlowDefinition.
 */
public final class MermaidGenerator {

    private MermaidGenerator() {}

    public static <S extends Enum<S> & FlowState> String generate(FlowDefinition<S> definition) {
        var sb = new StringBuilder();
        sb.append("stateDiagram-v2\n");

        S initial = definition.initialState();
        if (initial != null) sb.append("    [*] --> ").append(initial.name()).append('\n');

        Set<String> seen = new LinkedHashSet<>();
        for (Transition<S> t : definition.transitions()) {
            String key = t.from().name() + "->" + t.to().name();
            if (!seen.add(key)) continue;
            String label = transitionLabel(t);
            sb.append("    ").append(t.from().name()).append(" --> ").append(t.to().name());
            if (!label.isEmpty()) sb.append(" : ").append(label);
            sb.append('\n');
        }

        Set<String> errorSeen = new LinkedHashSet<>();
        for (var entry : definition.errorTransitions().entrySet()) {
            String key = entry.getKey().name() + "->" + entry.getValue().name();
            if (seen.contains(key)) continue;
            if (!errorSeen.add(key)) continue;
            sb.append("    ").append(entry.getKey().name())
              .append(" --> ").append(entry.getValue().name()).append(" : error\n");
        }

        for (S s : definition.terminalStates()) {
            sb.append("    ").append(s.name()).append(" --> [*]\n");
        }
        return sb.toString();
    }

    private static <S extends Enum<S> & FlowState> String transitionLabel(Transition<S> t) {
        return switch (t.type()) {
            case AUTO -> t.processor() != null ? t.processor().name() : "";
            case EXTERNAL -> t.guard() != null ? "[" + t.guard().name() + "]" : "";
            case BRANCH -> t.branch() != null ? t.branch().name() : "";
        };
    }

    /** Generate Mermaid data-flow diagram from requires/produces declarations. */
    public static <S extends Enum<S> & FlowState> String generateDataFlow(FlowDefinition<S> definition) {
        return definition.dataFlowGraph().toMermaid();
    }

    public static <S extends Enum<S> & FlowState> String writeToFile(
            FlowDefinition<S> definition, Path outputDir) throws IOException {
        Files.createDirectories(outputDir);
        String content = generate(definition);
        Files.writeString(outputDir.resolve("flow-" + definition.name() + ".mmd"), content);
        String dataFlow = generateDataFlow(definition);
        Files.writeString(outputDir.resolve("dataflow-" + definition.name() + ".mmd"), dataFlow);
        return content;
    }
}

File: ts/src/mermaid-generator.ts
import type { FlowDefinition } from './flow-definition.js';
import type { Transition } from './types.js';

export class MermaidGenerator {
  static generate<S extends string>(def: FlowDefinition<S>): string {
    const lines: string[] = ['stateDiagram-v2'];
    if (def.initialState) lines.push(`  [*] --> ${def.initialState}`);

    const seen = new Set<string>();
    for (const t of def.transitions) {
      const key = `${t.from}->${t.to}`;
      if (seen.has(key)) continue;
      seen.add(key);

      const label = this.transitionLabel(t);
      lines.push(label ? `  ${t.from} --> ${t.to}: ${label}` : `  ${t.from} --> ${t.to}`);
    }

    for (const [from, to] of def.errorTransitions) {
      const key = `${from}->${to}`;
      if (!seen.has(key)) {
        seen.add(key);
        lines.push(`  ${from} --> ${to}: error`);
      }
    }

    for (const s of def.terminalStates) {
      lines.push(`  ${s} --> [*]`);
    }

    return lines.join('\n') + '\n';
  }

  /** Generate Mermaid data-flow diagram from requires/produces declarations. */
  static generateDataFlow<S extends string>(def: FlowDefinition<S>): string {
    return def.dataFlowGraph?.toMermaid() ?? '';
  }

  private static transitionLabel<S extends string>(t: Transition<S>): string {
    if (t.type === 'auto') return t.processor?.name ?? '';
    if (t.type === 'external') return t.guard ? `[${t.guard.name}]` : '';
    if (t.type === 'branch') return t.branch?.name ?? '';
    return '';
  }
}
